### PR TITLE
KAFKA-2648: group.id is required for new consumer and cannot be empty

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 1]
```diff
@@ -381,8 +394,8 @@ public ConfigKey(String name, Type type, Object defaultValue, Validator validato
             this.defaultValue = defaultValue;
             this.validator = validator;
             this.importance = importance;
-            if (this.validator != null)
-                this.validator.ensureValid(name, defaultValue);
+            if (hasDefault() && validator != null)
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Why do we want to add this additional check?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: The validator checks don't make a lot of sense for the empty default value.

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Well, what about the validation such that the default value must be an empty string?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 2]
```diff
@@ -361,17 +361,10 @@ public void handle(JoinGroupResponse joinResponse, RequestFuture<ByteBuffer> fut
...
                         groupId);
                 future.raise(Errors.forCode(errorCode));
-            } else if (errorCode == Errors.INCONSISTENT_GROUP_PROTOCOL.code()
-                    || errorCode == Errors.INVALID_SESSION_TIMEOUT.code()) {
-                // log the error and re-throw the exception
+            } else {
                 Errors error = Errors.forCode(errorCode);
-                log.error("Attempt to join group {} failed due to: {}",
-                        groupId, error.exception().getMessage());
+                log.error("Attempt to join group {} failed due to: {}", groupId, error.exception().getMessage());
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Although theoretically we should not see any "unexpected error", I think it is a good sanity check moving forward if we changed the code but forget the update the error handling.

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Just to be clear, I think we can just add INVALID_GROUP_ID to be handled together with the other two, while keeping the unexpected error check.

**Comment by guozhangwang** [comment]
Thanks, LGTM.

**Pull request closed without merging**